src/board/v3/nap/track_channel.c

```diff
@@ -0,0 +1,152 @@
+                 - cp_start;
+  u32 samples = cp_units / cp_rate;
+  return samples;
+}
+
+void nap_track_init(u8 channel, gnss_signal_t sid, u32 ref_timing_count,
+                   float carrier_freq, float code_phase)
+{
+  u32 now = NAP->TIMING_COUNT;
+  u32 track_count = now + 200000;
+  double cp = propagate_code_phase(code_phase, carrier_freq,
+                                  track_count - ref_timing_count);
+
+  /* Contrive for the timing strobe to occur at or close to a
+   * PRN edge (code phase = 0) */
+  track_count += (SAMPLE_FREQ/GPS_CA_CHIPPING_RATE) * (1023.0-cp) *
+                 (1.0 + carrier_freq / GPS_L1_HZ);
+
+  u8 prn = sid.sat - 1;
+  NAP->TRK_CH[channel].CONTROL = prn << NAP_TRK_CONTROL_PRN_Pos;
+  /* We always start at zero code phase */
+  NAP->TRK_CH[channel].CODE_INIT_INT = 0;
+  NAP->TRK_CH[channel].CODE_INIT_FRAC = 0;
+  NAP->TRK_CH[channel].CODE_INIT_G1 = 0x3ff;
+  NAP->TRK_CH[channel].CODE_INIT_G2 = 0x3ff;
+
+  NAP->TRK_CH[channel].SPACING = (SPACING_HALF_CHIP << 16) | SPACING_HALF_CHIP;
+
+  u32 cp_rate = (1.0 + carrier_freq/GPS_L1_HZ) * GPS_CA_CHIPPING_RATE *
+                NAP_TRACK_CODE_PHASE_RATE_UNITS_PER_HZ;
+
+  NAP->TRK_CH[channel].CARR_PINC = -carrier_freq *
+                                   NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ;
+  NAP->TRK_CH[channel].CODE_PINC = cp_rate;
+  nap_ch_state[channel].len = NAP->TRK_CH[channel].LENGTH = calc_length_samples(1, 0, cp_rate)+1;
+  nap_ch_state[channel].code_phase = (NAP->TRK_CH[channel].LENGTH) * cp_rate;
+  NAP->TRK_CONTROL |= (1 << channel); /* Set to start on the timing strobe */
+
+  NAP->TRK_TIMING_COMPARE = track_count - SAMPLE_FREQ / GPS_CA_CHIPPING_RATE;
+
+  chThdSleep(CH_CFG_ST_FREQUENCY * ceil((float)(track_count - now)/SAMPLE_FREQ));
+}
+
+void nap_track_update(u8 channel, double carrier_freq,
+                      double code_phase_rate, u8 rollover_count,
+                      u8 corr_spacing)
+{
+  (void)corr_spacing; /* This is always written as 0 now... */
+
+  u32 cp_rate_fp = code_phase_rate * NAP_TRACK_CODE_PHASE_RATE_UNITS_PER_HZ;
+  NAP->TRK_CH[channel].CARR_PINC = -carrier_freq *
+                                   NAP_TRACK_CARRIER_FREQ_UNITS_PER_HZ;
+  NAP->TRK_CH[channel].CODE_PINC = cp_rate_fp;
+  NAP->TRK_CH[channel].LENGTH = calc_length_samples(rollover_count + 1,
+                                    nap_ch_state[channel].code_phase,
+                                    cp_rate_fp);
+
+  volatile u16 cp_int = NAP->TRK_CH[channel].CODE_PHASE_INT;
+  if ((cp_int != 0x3fe)) /* Sanity check, we should be just before rollover */
+    asm("nop");
+}
+
+void nap_track_read_results(u8 channel,
+                            u32* count_snapshot, corr_t corrs[],
+                            double *code_phase_early,
+                            double *carrier_phase)
+{
+  corr_t lc[5];
+  if (NAP->TRK_CH[channel].STATUS & 0x3F)
+    log_warn("Track correlator overflow 0x%08X on channel %d",
```

Comment:
We can remove it, but the scaling should be changed in the HDL so that it doesn't overflow.  I still think we'd like to know if it does.